File: python/voxgig_sekreto/plugins/onepassword.py
# The onepassword plugin: 1Password, through a Connect server. Needs
# HTTPS. A port of typescript/plugins/onepassword.ts.
from ..addr import checkaddr
from ..sekreto import SekretoError, checkname
from ..providers import Provider, providerplugin
from .httpjson import fetchjson, urlpart


class OnepasswordProvider(Provider):
    """1Password, through a Connect server.

    The item titled `api.token` (titles keep their dots), in the named
    vault. The value is the field with purpose PASSWORD, or the field
    labelled `value`. A vault that cannot be found is an error - config
    names it, so its absence is a broken store, not a missing secret.
    """

    def __init__(self, options=None):
        self.opts = options or {}
        self.vaultid = None

    def authheaders(self):
        return {'authorization': 'Bearer ' + (self.opts.get('token') or '')}

    def resolvevault(self, addr):
        want = self.opts.get('vault') or ''
        if '' == want:
            raise SekretoError('sekreto: onepassword: no vault')

        status, body = fetchjson('GET', addr + '/v1/vaults', self.authheaders())

        if 200 != status or not isinstance(body, list):
            raise SekretoError(
                'sekreto: onepassword error: ' + str(status) + ': listing vaults'
            )

        for entry in body:
            if isinstance(entry, dict) and (want == entry.get('id') or want == entry.get('name')):
                return str(entry.get('id'))

        raise SekretoError('sekreto: onepassword: no vault named ' + want)

    def lookup(self, name):
        checkname(name)

        addr = (self.opts.get('addr') or '').rstrip('/')
        if '' == addr:
            raise SekretoError('sekreto: onepassword: no addr')
        checkaddr(addr)

        if self.vaultid is None:
            self.vaultid = self.resolvevault(addr)

        filterpart = urlpart('title eq "' + name + '"')
        status, found = fetchjson(
            'GET',
            addr + '/v1/vaults/' + self.vaultid + '/items?filter=' + filterpart,
            self.authheaders(),
        )

        if 200 != status or not isinstance(found, list):
            raise SekretoError(
                'sekreto: onepassword error: ' + str(status) + ': finding ' + name
            )

        if 0 == len(found):
            return None

        status, item = fetchjson(
            'GET',
            addr + '/v1/vaults/' + self.vaultid + '/items/' + str(found[0].get('id')),
            self.authheaders(),
        )

        if 200 != status:
            raise SekretoError(
                'sekreto: onepassword error: ' + str(status) + ': reading ' + name
            )

        fields = (item or {}).get('fields') or []

        for field in fields:
            if isinstance(field, dict) and 'PASSWORD' == field.get('purpose'):
                value = field.get('value')
                return None if value is None else str(value)
        for field in fields:
            if isinstance(field, dict) and 'value' == field.get('label'):
                value = field.get('value')
                return None if value is None else str(value)

        return None

    def describe(self):
        return 'onepassword:' + (self.opts.get('vault') or '')


# The plugin: the `onepassword` provider kind, as a voxgig/plugin definition.
onepassword = providerplugin('onepassword', OnepasswordProvider)

File: ruby/lib/voxgig_sekreto/plugins/onepassword.rb
# frozen_string_literal: true

# The onepassword plugin: 1Password, through a Connect server. Needs
# HTTPS. A port of typescript/plugins/onepassword.ts, which is canonical.

require_relative '../../voxgig_sekreto'
require_relative '../addr'
require_relative 'httpjson'

module VoxgigSekreto
  # 1Password, through a Connect server.
  #
  # The item titled `api.token` (titles keep their dots), in the named
  # vault. The value is the field with purpose PASSWORD, or the field
  # labelled `value`. A vault that cannot be found is an error - config
  # names it, so its absence is a broken store, not a missing secret.
  class OnepasswordProvider
    def initialize(opts = nil)
      @opts = opts || {}
      @vaultid = nil
    end

    def auth
      { 'authorization' => 'Bearer ' + (@opts['token'] || '') }
    end

    def resolvevault(addr)
      want = @opts['vault'] || ''
      raise SekretoError, 'sekreto: onepassword: no vault' if '' == want

      status, body = VoxgigSekreto.fetchjson('GET', addr + '/v1/vaults', auth)

      unless 200 == status && body.is_a?(Array)
        raise SekretoError, 'sekreto: onepassword error: ' + status.to_s + ': listing vaults'
      end

      body.each do |entry|
        return entry['id'].to_s if entry.is_a?(Hash) && (want == entry['id'] || want == entry['name'])
      end

      raise SekretoError, 'sekreto: onepassword: no vault named ' + want
    end

    def lookup(name)
      VoxgigSekreto.checkname(name)

      addr = (@opts['addr'] || '').sub(%r{/\z}, '')
      raise SekretoError, 'sekreto: onepassword: no addr' if '' == addr

      VoxgigSekreto.checkaddr(addr)

      @vaultid = resolvevault(addr) if @vaultid.nil?

      filter = VoxgigSekreto.uriescape('title eq "' + name + '"')
      status, found = VoxgigSekreto.fetchjson(
        'GET', addr + '/v1/vaults/' + @vaultid + '/items?filter=' + filter, auth
      )

      unless 200 == status && found.is_a?(Array)
        raise SekretoError, 'sekreto: onepassword error: ' + status.to_s + ': finding ' + name
      end

      return nil if found.empty?

      status, item = VoxgigSekreto.fetchjson(
        'GET', addr + '/v1/vaults/' + @vaultid + '/items/' + found[0]['id'].to_s, auth
      )

      if 200 != status
        raise SekretoError, 'sekreto: onepassword error: ' + status.to_s + ': reading ' + name
      end

      fields = item.is_a?(Hash) && item['fields'].is_a?(Array) ? item['fields'] : []

      fields.each do |field|
        next unless field.is_a?(Hash) && 'PASSWORD' == field['purpose']

        return field['value'].nil? ? nil : field['value'].to_s
      end
      fields.each do |field|
        next unless field.is_a?(Hash) && 'value' == field['label']

        return field['value'].nil? ? nil : field['value'].to_s
      end

      nil
    end

    def describe
      'onepassword:' + (@opts['vault'] || '')
    end
  end

  module Plugins
    # The plugin: the `onepassword` provider kind, as a voxgig/plugin
    # definition.
    ONEPASSWORD = VoxgigSekreto.providerplugin('onepassword', lambda { |spec|
      OnepasswordProvider.new(spec.slice('addr', 'token', 'vault'))
    })
  end
end
